RetroPlayer: Let ApplyStateBlock() handle the VAO rebind

KOPRajs opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
This PR fixes following issues in RetroPlayer rendering:
- Invalid OpenGL state after VAO unbind, see here: https://github.com/xbmc/xbmc/pull/26435#discussion_r2116610809
- Restoring rendering state without saving it first, see here: https://github.com/xbmc/xbmc/pull/26435#issuecomment-2965471282
- Some minor cleanups, removed commented code, renamed variables for consistency etc.

I have kept the work split into 3 commits, even though the last one partially reverts the previous one, because it seems more understandable this way and the code without the last commit is fully functional as well.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->


## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I've only tested it on Linux (Wayland/GL, Wayland/GLES, GBM/GL and GBM/GLES). It needs to be tested on all platforms including Windows before merge. Maybe @garbear can include this in his RP test builds?

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
I hope it can fix remaining Kodi GUI corruption issues seen after using RetroPlayer.


## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

KOPRajs committed:
RetroPlayer: Rename GL shader array buffers for consistency

KOPRajs commented on xbmc/cores/RetroPlayer/rendering/VideoRenderers/RPRendererDMAOpenGLES.cpp:
```diff
@@ -106,84 +107,85 @@ void CRPRendererDMAOpenGLES::Render(uint8_t alpha)
-  glVertexAttribPointer(loc, 2, GL_FLOAT, 0, sizeof(PackedVertex),
-                        reinterpret_cast<const GLuint*>(offsetof(PackedVertex, u1)));
+    glVertexAttribPointer(vertLoc, 3, GL_FLOAT, 0, sizeof(PackedVertex),
+                          reinterpret_cast<const GLvoid*>(offsetof(PackedVertex, x)));
```
I've noticed that there was _GLuint*_ instead of _GLvoid*_ here. The function accepts _GLvoid*_ and there is _GLvoid*_ in the RPRendererOpenGLES.cpp (non DMA), so I've changed it here to the same.

garbear commented:
When bisecting, it seems to be the second commit where the corruption starts to occur.

KOPRajs commented:
> Tested on macOS ARM. I got some texture corruption. Also a few shaders are broken:
> 
>     * xBR-lv2
>     * xBR-lv2 (no blend)
>     * xBR-lv3
>     * xBR-mlv4
>     * Amiga A520
>     * GBC 4x

Thanks for testing. Can you provide screenshots? Is GUI corrupted, if you open it during the game (press _Home_ while in game)? Or does the corruption come after exiting? What is the version and OpenGL profile used by the macOS? Is it running using OpenGL 4.6 Core profile?

KOPRajs commented:
According to this (https://stackoverflow.com/questions/26552642/when-is-what-bound-to-a-vao) and some other references it seems that the _GL_ARRAY_BUFFER_ is not part of the VAO state after all, so we are supposed to unbind it.

I've added it back along with a minor change in order (which should not have any impact), so please test it again when you can (but I don't expect any change to be honest).

I've tested the current version on all 4 Linux builds again and all shaders work without any issues.

Please note, that all shaders work means, that _xBR-lv2_, _xBR-lv2 (no blend)_, _Amiga A520_ and _PAL_ should not work for OpenGL ES. This is caused by the shaders themselves and it is expected. These 4 only supports OpenGL Core Profile.

Unfortunately, I'm going offline for about 2 weeks now, so I'll continue with this when I come back.

sarbes commented:
Looking at the code, I don't think you are on the right track. The way is it is now, neither VAOs, VBOs or IBOs are used properly. They might not even exist to begin with.

You need to break it down into multiple parts:

OnInit():
- create VBO
- bind VBO 
- upload VBO
- create VAO
- bind VAO
- vertex attribs
- create IBO
- bind IBO
- upload IBO
- unbind VAO
- unbind VBO

OnResolutionChange()
- bind VBO
- upload VBO
- bind VAO
- vertex attribs
- unbind VAO
- unbind VBO

OnRender()
- bind VAO
- setup uniforms
- draw 
- bind default VAO

garbear commented:
Hm, I was testing _without_ this PR, and the corruption started happening against master. So this PR doesn't seem to be the cause of the corruption.

Here's what it looks like.

<img width="1274" alt="Screenshot 2025-01-15 at 10 25 43 AM" src="https://github.com/user-attachments/assets/86c052a2-18c7-4ac2-acf2-5eabf0335cdb" />

KOPRajs commented:
> Hm, I was testing _without_ this PR, and the corruption started happening against master. So this PR doesn't seem to be the cause of the corruption.

Can you re-test the version, which was the last known to work on macOS? Does it still work? It might be unrelated to the VAO after all.

garbear commented:
> Can you re-test the version, which was the last known to work on macOS? Does it still work? It might be unrelated to the VAO after all.

It does not work, so likely unrelated.

> EDIT: It seems to be the same picture. Does the corruption always look exactly the same?

It changes, it happens immediately on the home screen (or maybe after flipping through a few labs on the left). The home screen is always what I return to and in that sense it's similar because the same window is observed. But I'll edit this post with more pics to see if it gives us a clue.

KOPRajs committed:
RetroPlayer: Let ApplyStateBlock() handle the VAO rebind

KOPRajs commented:
Here is a final summary of the changes:

1. Removed unbinding of EBO added here: https://github.com/xbmc/xbmc/pull/26435#issuecomment-2920271213
   - it is not correct, see here: https://github.com/xbmc/xbmc/pull/26435#discussion_r2116610809
   - and it doesn't fix anything

2. Added _CaptureStateBlock()_ to _PreRender()_
   - the original code was using _ApplyStateBlock()_ in _PostRender()_ without saving the state first

3. Moved _PreRender()_ to _RenderFrame()_
   - if the _m_renderBuffer_ was null, the _PreRender()_ was still called, but the rest (_RenderInternal()_, _PostRender()_) was aborted, this change ensures that the _PreRender()_ is always followed by _RenderInternal()_ and _PostRender()_

4. Added _CaptureStateBlock()_ and _ApplyStateBlock()_ to the GL renderer initialization
   - we need _ApplyStateBlock()_ to restore the global VAO after setting up of the renderer VAO and the black bars VAO here
   - added the same to the GLES renderer as well, even though it is not needed there, to maintain consistency of the code structure

5. Removed _ApplyStateBlock()_ form the _CShaderPreset_ destructor
   - the object management is not related to the rendering state
   - we don't know when this is going to be called
   - we haven't captured the state before

6. Added VAO initialization to _CShaderGL::Create()_ and removed EBO binding from _CShaderGL::Render()_
   - the EBO is part of the VAO state, so we should only bind it once

7. Removed _glDisableVertexAttribArray()_ from _CShaderGL::Render()_
   - these are part of the VAO as well, so no need to disable them here
   - they are not disabled in the GUI shader rendering code as well, so let's be consistent

8. Changed the format of the condition for rendering via GUI shaders or via custom shader preset to _if_ and _else_
   - this change makes the diff bigger and less understandable, but the code is better
   - the original condition was _if (m_bUseShaderPreset) {break;} do the opposite_
   - now it is _if (m_bUseShaderPreset) {} else {}_
   - this makes it possible to add a common code after the block
   - it is consistent with the Windows version
   - I expect that this was previously changed to make the original PR smaller and easier to read, but we can afford to change it back now

garbear commented:
Copilot issues are benign, are you happy to merge?
